kratos/solving_strategies/strategies/explicit_solving_strategy_bfecc.h

```diff
@@ -0,0 +1,546 @@
+                time_integration_theta = 1.0;
+                time_direction =-1.0;
+                break;
+            case FINAL:
+                InitializeBFECCFinalSubstep();
+                time_integration_theta = 0.0;
+                time_direction = 1.0;
+                break;
+            default:
+                KRATOS_ERROR << "Invalid value for Substep" << std::endl;
+        }
+
+        r_process_info.GetValue(TIME_INTEGRATION_THETA) = time_integration_theta;
+        explicit_bs.BuildRHS(r_model_part);
+
+        IndexPartition<SizeType>(r_dof_set.size()).for_each(
+            [&](SizeType i_dof){
+                auto it_dof = r_dof_set.begin() + i_dof;
+
+                // Save current value in the corresponding vector
+                const double residual = it_dof->GetSolutionStepReactionValue();
+
+                // Do the DOF update
+                double& r_u = it_dof->GetSolutionStepValue(0);
+                double& r_u_prev_step = it_dof->GetSolutionStepValue(1);
+
+                if (!it_dof->IsFixed()) {
+                    const double mass = r_lumped_mass_vector[i_dof];
+                    r_u += time_direction * (dt / mass) * residual;
+                } else {
+                    r_u = time_integration_theta*rFixedDofsValues[i_dof] + (1 - time_integration_theta)*r_u_prev_step;
+                }
+            }
+        );
+
+        switch(SubstepType) {
+            case FORWARD:  FinalizeBFECCForwardSubstep();   break;
+            case BACKWARD: FinalizeBFECCBackwardSubstep();  break;
+            case FINAL:    FinalizeBFECCFinalSubstep();     break;
+            default: KRATOS_ERROR << "Invalid value for Substep" << std::endl;
+        }
+
+        KRATOS_CATCH("Substep type: " + [](Substep substep) -> std::string {
+            switch(substep) {
+            case FORWARD:  return "FORWARD";
+            case BACKWARD: return "BACKWARD";
+            case FINAL:    return "FINAL";
+            default: return std::to_string(static_cast<int>(substep));
+            }
+        }(SubstepType));
+    }
+
+    /**
+     * Computes the error by comparing the buffer position 0 and the previous step
+     * solution. Then stores the corrected starting point in the buffer position 1.
+     * @param rPrevStepSolution
+     */
+    void CorrectErrorAfterForwardsAndBackwards(const LocalSystemVectorType& rPrevStepSolution)
+    {
+        auto& explicit_bs = BaseType::GetExplicitBuilder();
+        auto& r_dof_set = explicit_bs.GetDofSet();
+
+        IndexPartition<SizeType>(r_dof_set.size()).for_each(
+            [&](const SizeType dof_index)
+            {
+                const double prev_step_solution = rPrevStepSolution[dof_index];
+                auto& r_dof = *(r_dof_set.begin() + dof_index);
+
+                const double error = (r_dof.GetSolutionStepValue(0) - prev_step_solution)/2.0;
+
+                r_dof.GetSolutionStepValue(1) = prev_step_solution - error;
```

Comment:
Maybe I'm missing something but shouldn't the error correction be applied in the current step value?

Reply:
It must be applied to both. Here it is applied only to (1), because afterwards in line 286 it is copied into (0) during CopySolutionStepData (which must be called anyways to extract the fixed dofs).

Reply:
OK. Then I'd put a comment explaining why it is not copied to the current position of the database.